codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -214,7 +232,99 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+
+        // We need at least 4 byte for a valid TLV. If we have less bytes available, we need to ignore
+        if (header.readableBytes() < 4) {
+            return Collections.emptyList();
+        }
+
+        // In most cases there are less than 4 TLVs available
+        final List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        while (header.isReadable()) {
+            final HAProxyTLV haProxyTLV = readNextTLV(header);
+
+            if (haProxyTLV == null) {
+                // We don't have a TLV available anymore, we can stop
+                break;
+            }
+            haProxyTLVs.add(haProxyTLV);
+            if (haProxyTLV instanceof HAProxySSLTLV) {
+                haProxyTLVs.addAll(((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs());
+            }
+        }
+
+        return haProxyTLVs;
+    }
+
+    private static HAProxyTLV readNextTLV(final ByteBuf header) {
+
+        // We need at least 4 bytes for a TLV
+        if (header.readableBytes() < 4) {
+            return null;
+        }
+
+        final byte typeAsByte = header.readByte();
+        final HAProxyTLV.Type type = HAProxyTLV.Type.typeForByteValue(typeAsByte);
+
+        final int length = header.readUnsignedShort();
+        switch (type) {
+        case PP2_TYPE_SSL:
+            final byte[] rawContent = copyByteBufToByteArray(header, length);
```

Comment:
The intention was that the HAProxyTLV may be used by business logic and thus I see two reasons for the byte array: 

1. The can reside in memory where it's outside of Netty's control. This has the advantage that no reference counting needs to be in place and the TLV values can be in memory as long as needed by the business application (in our use case the SSL information is actually in memory all the time for each channel!)
2. While the ByteBuf API is incredibly useful for the networking layer, it makes consuming "plain ol'" bytes harder for the business layer (which in our case expects byte arrays), because of the distinction between on- and off-heap memory. So the main argument here would be usability at the cost of a copy operation.

If the copy operation would be a show-stopper, I can certainly use ByteBufs here instead